File: cs/Problems/Day24.cs
namespace AOC2023.Problems;

public class Day24 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => CountHailIntersections(input);

    private static int CountHailIntersections(ReadOnlySpan<char> input)
    {
        Span<Range> lineRanges = stackalloc Range[301];
        int lines = input.Split(lineRanges, InputReader.NewLine);

        var firstLine = input[lineRanges[0]];
        long minTime = long.Parse(firstLine[..firstLine.IndexOf(' ')]);
        long maxTime = long.Parse(firstLine[(firstLine.IndexOf(' ') + 1)..]);

        Span<Hailstone> hailStones = stackalloc Hailstone[lines];

        for (int i = 1; i < lines; i++)
            hailStones[i] = Hailstone.Parse(i, input[lineRanges[i]]);

        int intersects = 0;

        for (int i = 1; i < lines - 1; i++)
        for (int j = i + 1; j < lines; j++)
            intersects += hailStones[i].Intersects(hailStones[j], minTime, maxTime) ? 1 : 0;

        return intersects;
    }

    readonly record struct Hailstone(int Id, Point Position, Velocity Velocity)
    {
        public readonly bool Intersects(Hailstone other, long minTime, long maxTime)
        {
            // Intersection of dimensional lines formula, found it somewhere else.
            decimal determinant = Velocity.VX * other.Velocity.VY - Velocity.VY * other.Velocity.VX;

            if (determinant == 0)
                return false;

            decimal t1 = Velocity.VX * Position.Y - Velocity.VY * Position.X;
            decimal t2 = other.Velocity.VX * other.Position.Y - other.Velocity.VY * other.Position.X;

            decimal xIntersect = (other.Velocity.VX * t1 - Velocity.VX * t2) / determinant;
            decimal yIntersect = (other.Velocity.VY * t1 - Velocity.VY * t2) / determinant;

            if (xIntersect < minTime || xIntersect > maxTime || yIntersect < minTime || yIntersect > maxTime)
                return false;

            bool thisIntersect = Math.Sign(xIntersect - Position.X) != Math.Sign(Velocity.VX);
            bool otherIntersect = Math.Sign(xIntersect - other.Position.X) != Math.Sign(other.Velocity.VX);

            if (thisIntersect || otherIntersect)
                return false;

            return true;
        }

        public static Hailstone Parse(int id, ReadOnlySpan<char> hailString)
        {
            int separatorInx = hailString.IndexOf(" @ ");
            var positions = hailString[..separatorInx];
            var velocities = hailString[(separatorInx + 3)..];

            Span<long> position = stackalloc long[3];
            Span<int> velocity = stackalloc int[3];

            InputParser.ParseNumbers(ref position, positions, 3, ", ");
            InputParser.ParseNumbers(ref velocity, velocities, 3, ", ");

            return new Hailstone(id, new(position[0], position[1]), new(velocity[0], velocity[1]));
        }
    }

    readonly record struct Point(long X, long Y);

    readonly record struct Velocity(int VX, int VY);
}

File: ts/problems/day24.ts
import { NEW_LINE } from "../inputReader.js";

const countHailIntersections = (input: string) => {
  const [firstLine, ...hailStrings] = input.split(NEW_LINE);
  const [minTime, maxTime] = firstLine.split(" ").map(Number);

  return hailStrings
    .map(parseHailstone)
    .reduce((acc, _, i, hailStones) => acc + countNextIntersections(hailStones, i, minTime, maxTime), 0);
};

const countNextIntersections = (hailStones: Hailstone[], from: number, minTime: number, maxTime: number) =>
  hailStones
    .slice(from + 1)
    .reduce((acc, curr) => acc + Number(hailstonesIntersect(hailStones[from], curr, minTime, maxTime)), 0);

const parseHailstone = (hailString: string): Hailstone => {
  const [positions, velocities] = hailString.split(" @ ");
  const [x, y] = positions.split(", ").map(Number);
  const [vx, vy] = velocities.split(", ").map(Number);

  return { position: { x, y }, velocity: { vx, vy } };
};

const hailstonesIntersect = (first: Hailstone, second: Hailstone, minTime: number, maxTime: number) => {
  if (first === second) return false; // Can't intersect themselves.

  // Intersection of dimensional lines formula, found it somewhere else.
  const determinant = first.velocity.vx * second.velocity.vy - first.velocity.vy * second.velocity.vx;

  if (determinant == 0) return false;

  const t1 = first.velocity.vx * first.position.y - first.velocity.vy * first.position.x;
  const t2 = second.velocity.vx * second.position.y - second.velocity.vy * second.position.x;

  const xIntersect = (second.velocity.vx * t1 - first.velocity.vx * t2) / determinant;
  const yIntersect = (second.velocity.vy * t1 - first.velocity.vy * t2) / determinant;

  if (xIntersect < minTime || xIntersect > maxTime || yIntersect < minTime || yIntersect > maxTime) return false;

  const thisIntersect = Math.sign(xIntersect - first.position.x) != Math.sign(first.velocity.vx);
  const otherIntersect = Math.sign(xIntersect - second.position.x) != Math.sign(second.velocity.vx);

  if (thisIntersect || otherIntersect) return false;

  return true;
};

type Hailstone = { position: Point; velocity: Velocity };

type Point = { x: number; y: number };

type Velocity = { vx: number; vy: number };

export default countHailIntersections;
